src/mode/mode.ts

```diff
@@ -10,14 +10,34 @@ export enum Mode {
   Visual,
   VisualBlock,
   VisualLine,
+  Select,
+  SelectBlock,
+  SelectLine,
   SearchInProgressMode,
   CommandlineInProgress,
   Replace,
   EasyMotionMode,
   EasyMotionInputMode,
   SurroundInputMode,
-  OperatorPendingMode, // Pseudo-Mode, used only when remapping. DON'T SET TO THIS MODE
   Disabled,
+  // The following modes are Pseudo-Modes, used only when remapping or for 'showmode'
```

Comment:
I wonder if there's a better way of encoding this information, rather than an ever-growing list of modes. Any thoughts?

Reply:
We could do something like this:
```ts
export enum DefaultMode {
  Normal,
  Insert,
  Visual,
  VisualBlock,
  VisualLine,
  Select,
  SelectBlock,
  SelectLine,
  SearchInProgressMode,
  CommandlineInProgress,
  Replace,
  Disabled,
}

export enum PluginMode {
  EasyMotionMode,
  EasyMotionInputMode,
  SurroundInputMode,
}

export enum PseudoMode {
  // The following modes are Pseudo-Modes, used only when remapping or for 'showmode'
  // to give feedback to user.
  // DON'T SET TO THESE MODES!!!
  OperatorPendingMode,
  InsertNormal,
  ReplaceNormal,
  InsertVisual,
  InsertVisualBlock,
  InsertVisualLine,
  ReplaceVisual,
  ReplaceVisualBlock,
  ReplaceVisualLine,
  InsertSelect,
  InsertSelectBlock,
  InsertSelectLine,
  ReplaceSelect,
  ReplaceSelectBlock,
  ReplaceSelectLine,
}

export type Mode = DefaultMode | PluginMode | PseudoMode;
export const Mode = { ...DefaultMode, ...PluginMode, ...PseudoMode };

```

This way we could separate the modes and everything else should still work the same. Then on functions like `setCurrentMode(mode : Mode)` we would change it to `setCurrentMode(mode : DefaultMode | PluginMode)` this way you could never set to a pseudo-mode.

The only issue with this approach would be that all pieces of code like this:
```ts
if ([Mode.Insert, Mode.Replace].includes(vimState.currentMode))
```
would have to be changed to:
```ts
if (([Mode.Insert, Mode.Replace] as Mode[]).includes(vimState.currentMode))
```

Reply:
That seems slightly better to me, but still not ideal. I think this deserves more thought; let's keep it as is for now.